velox/dwio/parquet/thrift/ThriftStreamingTransport.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <thrift/transport/TVirtualTransport.h>
+#include "velox/dwio/common/BufferedInput.h"
+#include "velox/dwio/parquet/thrift/ThriftTransport.h"
+
+namespace facebook::velox::parquet::thrift {
+
+class ThriftStreamingTransport : public ThriftTransport {
+ public:
+  ThriftStreamingTransport(
+      std::shared_ptr<dwio::common::SeekableInputStream> inputStream,
+      const char* FOLLY_NULLABLE& bufferStart,
+      const char* FOLLY_NULLABLE& bufferEnd)
+      : ThriftTransport(),
+        inputStream_(inputStream),
+        bufferStart_(bufferStart),
+        bufferEnd_(bufferEnd) {
+    VELOX_CHECK_NOT_NULL(inputStream_);
+    VELOX_CHECK_NOT_NULL(bufferStart_);
+    VELOX_CHECK_NOT_NULL(bufferEnd_);
+  }
+
+  uint32_t read(uint8_t* outputBuf, uint32_t len) {
```

Comment:
Also thinking about how to avoid copy in cases when possible.  Maybe the signature should be

```
uint8_t* read(uint8_t* outputBuf, uint32_t len, uint32_t& outLen)
```

Where we return `bufferStart_` if `bufferSize_ >= len`, or we return `outputBuf` if copy is needed.  `outLen` should be the actual size of the buffer returned.

Reply:
In this case maybe we can even manage `outputBuf` as a class variable ourselves

Reply:
I see your point. This `ThriftTransport` class is designed to optimize the Thrift generated code. For example, there are functions like `readI32()` in the generated code. Originally, these Thrift calls would trigger a `pread()` on the file, which is super expensive. This `ThriftTransport` is to overload the `read()` and avoid `pread()`. So the function signature should be aligned with the Thrift generated code. 

Also, the unnecessary deep copy is from `readPageHeader()`
`if (bufferEnd_ - bufferStart_ >= sizeof(PageHeader)) {
    wasInBuffer = true;
    transport = std::make_shared<thrift::ThriftBufferedTransport>(
        bufferStart_, sizeof(PageHeader));
    protocol = std::make_unique<apache::thrift::protocol::TCompactProtocolT<
        thrift::ThriftBufferedTransport>>(transport);
  } else {
    dwio::common::readBytes(
        std::min<int64_t>(remainingSize, sizeof(PageHeader)),
        inputStream_.get(),
        &copy,
        bufferStart_,
        bufferEnd_);

    transport = std::make_shared<thrift::ThriftBufferedTransport>(
        copy, sizeof(PageHeader));`

Reply:
So you are planning to keep the first branch of `if` to handle the case where we don't need copy?

Reply:
Yes, if it hits the first branch, there is no deep copy. This will be updated in my next PR. My next PR will refactor this `readPageHeader()` . The `if-else` statement can be removed. The input is just the `inputStream_`. This is able remove the deep copy in else section and fix the incorrect page header length.

Reply:
And within the scope of this PR, the logic of `readPageHeader()` remains unchanged. This PR is the prerequisite for the updated  https://github.com/facebookincubator/velox/pull/4108.